### ISSUE TITLE
Hypre comm package not cleanly deleted after `HypreParMatrix * Add()`

### ISSUE THREAD

**Issue by barker29** [mfem/mfem#2035]
The `HypreParMatrix` returned from `HypreParMatrix * Add(double, HypreParMatrix*, double, HypreParMatrix*)` does not cleanly delete its `CommPkg` when using hypre 2.19.0 (and probably other versions).

To reproduce this issue, run the following code through valgrind:
```c++
#include "mfem.hpp"
using namespace mfem;

int main(int argc, char *argv[])
{
   MPI_Init(&argc, &argv);
   const char *mesh_file = "../data/star.mesh";
   Mesh mesh(mesh_file, 1, 1);
   int dim = mesh.Dimension();
   ParMesh pmesh(MPI_COMM_WORLD, mesh);

   FiniteElementCollection *fec = new H1_FECollection(1, dim);
   ParFiniteElementSpace fespace(&pmesh, fec);

   ParBilinearForm a(&fespace);
   a.AddDomainIntegrator(new DiffusionIntegrator);
   a.Assemble();
   OperatorPtr A;
   Array<int> empty;
   a.FormSystemMatrix(empty, A);

   HypreParMatrix * hA = A.As<HypreParMatrix>();
   {
      HypreParMatrix * hAT = hA->Transpose();
      HypreParMatrix * yet_another = Add(1.0, *hA, 0.5, *hAT);
      delete hAT;
      delete yet_another;
   }

   delete fec;
   MPI_Finalize();
   return 0;
}
```
If the part in inner braces is removed, this runs valgrind clean. The valgrind trace points to the `Add` function, not the `Transpose()`.

**Comment by tzanio** [comment]
I don't see anything suspicious in the [`Add`](https://github.com/mfem/mfem/blob/master/linalg/hypre.cpp#L1690-L1698) function or in [`hypre_ParCSRMatrixAdd`](https://github.com/mfem/mfem/blob/master/linalg/hypre_parcsr.cpp#L1757). 

Hopefully this is not due to some change in the handling of `CommPkg`... we have that in quite a few places.

Just to check -- do you see this issue with `ParAdd` and `ParMult`?

**Comment by barker29** [comment]
I see exactly the same issue with `ParAdd` but not `ParMult`, `ParMult` runs valgrind clean. I've also duplicated the same behavior with hypre 2.16.0, I'll try some other hypre versions later on.

**Comment by tzanio** [comment]
This may be related to https://github.com/mfem/mfem/issues/1750

**Comment by v-dobrev** [comment]
Adding `bug` label to prevent the "stale bot" from marking the issue as stale (I think the `todo` label does not do that at the moment).

**Comment by tzanio** [comment]
@barker29, were you able to get any additional information on this?

**Comment by barker29** [comment]
For `Add` as in the code snippet above, it runs valgrind clean up to and including hypre v2.13.0 and then you see memory leaks in v2.14.0 (I only looked at tagged versions of hypre, didn't try to go finer grained than that). The valgrind details point to the `CommPkg` as not being deleted.

For `ParMult`, everything is fine up to and including v2.18.2 and then you get "still reachable" memory in v2.19.0, which is probably not as serious a concern as the "definitely lost" memory in `Add`. I thought at first this was also the `CommPkg` but now I think it's something else.

**Comment by v-dobrev** [comment]
Maybe the issue with `ParMult` is related to the new `HYPRE_Finalize()` which I think was added in v2.19.0.

For, `Add` we can try the valgrind option `--track-origins=yes` to see where the leaked memory was allocated -- that may give us some clue.

**Comment by barker29** [comment]
The issue with `ParMult` is indeed related to `HYPRE_Init()`, `HYPRE_Finalize()` (see #1276 , #1750).

The issue with `Add`, `ParAdd` is slightly different and is fixed in #2147

**Comment by tzanio** [comment]
Discussion continued in #2147.
